Upgrade the version of the dependency controller-runtime from `v0.1.10` to `v0.2.0-beta.1`

camilamacedo86 opened the pull request:
**Description of the change:**

### Changed
- Upgrade the version of the dependency [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) from `v0.1.10` to `v0.2.0-beta.1`
- Upgrade Kubernetes dependency versions from `kubernetes-1.13.1` to `kubernetes-1.14.1`
- Upgrade github.com/operator-framework/operator-lifecycle-manager dependency from the commit `b8a4faf68e36feb6d99a6aec623b405e587b17b1` to `tag 0.10.1`
- Upgrade k8s.io/kube-openapi from `0cf8f7e6ed1d2e3d47d02e3b6e559369af24d8`  dependency to a `01b7d5d6c2258c80a4a10070f3dee9cd575d9c7`
- Upgrade others menor versions. 

**Motivation for the change:**
Address the commit https://github.com/kubernetes-sigs/controller-runtime/commit/fc804a411c06919561edbcb0453601c436556613

- https://github.com/operator-framework/operator-sdk/issues/1386
- Allow users to watch/cache N namespaces as follows. 

```go
         // Create a new Cmd to provide shared dependencies and start components
	mgr, err := manager.New(cfg, manager.Options{
		NewCache: cache.MultiNamespacedCacheBuilder(namespaces),
	})
```

## Progress
Working on in Rebase/Merge the changes/fixes from @estroz #1512

camilamacedo86 commented:
/retest

camilamacedo86 commented:
Shows that all is working fine but it has some issue with the env. I do not have permission to try to run just the goal again. 

Following the error faced:
 ```
       	wait_util.go:64: Waiting for full availability of example-memcached deployment (0/4)
        	wait_util.go:64: Waiting for full availability of example-memcached deployment (0/4)
        	memcached_test.go:123: timed out waiting for the condition
```
<img width="622" alt="Screenshot 2019-05-06 at 01 01 27" src="https://user-images.githubusercontent.com/7708031/57202233-a5dece80-6f9a-11e9-8733-ad695237b300.png">

camilamacedo86 commented:
Hi @LiliC,

I checked the PR https://github.com/operator-framework/operator-sdk/pull/1369 made by @corinnekrych. This PR shows missing changes and because of it is not passing the ansible as well. Also, she increases the version of Kube. 

This one here, shows be working very well but the go test is facing timeout which I am not sure if is related or not with the changes performed. It is something that I need a help from you folks maintainers from @LiliC operator-sdk. Is the timeout issue in the go test faced usual? Could we try to just re-run this one? 

@corinnekrych let's combine the forces? I will ping you tomorrow for we check how we can move together.

estroz commented:
I'd rather wait for a v0.2.0 stable release. There's no urgency to upgrade for the next release.

camilamacedo86 commented:
@estroz,

Unfortunately for us has urgency since we are stuck and not able to attend this requirement of our projects which are using it. We cannot allow the operator to watch/cache other namespaces just the specific ones and because of this huge block in our side, I decided to contribute with this PR. 

c/c @LiliC @shawn

LiliC commented:
/ok-to-test

LiliC commented:
> Unfortunately for us has urgency since we are stuck for not being able to attend a requirement in our projects which are using it.

Sorry to hear that, but we can't have a nonstable release of such a huge dependency in our project, as we do promise certain stability to our users. @camilamacedo86 we discussed this and we want to wait until Wednesday at least, we will know when the stable controller-runtime release will happen and we will get back to you then. But we would be happy to have this contribution!

LiliC commented:
A couple of things before this is ready to be merged:

- [ ] stable release of controller-runtime
- [ ] migration document as it has a lot of breaking changes for users using current 0.7.0 release
- [ ] we might want to release one more time before this gets merged

cc @joelanford @theishshah @shawn-hurley @hasbro17 @AlexNPavel

camilamacedo86 commented:
@LiliC Wednesday or 1 or 2 weeks is 100% fine we just cannot wait for months to get it.

camilamacedo86 commented:
Hi @LiliC, 

Now all shows working fine here. 

The changes that need to be done by in the user's project from the latest version to this one is in the section break changes in the Changelog. See https://github.com/operator-framework/operator-sdk/commit/def0660143f2b755955bda869461d1938c74ab72

You added `migration document as it has a lot of breaking changes for users using current 0.7.0 release` , however, for the projects which are using 0.7.0(latest) to this one are not too many changes and the current migration doc is not covering from  0.1.0 to 0.7.0 either. So, shows that do the migration doc from 0.1.0 to 0.7.0 is not part of this task. Could we leave it to be done in another PR as part of another task?

LiliC commented:
@camilamacedo86 Yes we have an issue open to update the migration guide, we put breaking changes in the CHANGELOG. And besides new files and commands added, there were not many major changes since 0.1.0 onwards but yes we need to update that in the near future. 

We can help with adding the migration guide no problem regarding that, but we agreed to not merge this until we add that to the PR cc @hasbro17 ^

The new stable release of controller-runtime will be before kubecon which is in 10 days time or so, according to yesterdays SIG meeting, so we should be able to merge that then. If this is delaying your progress on the operator you can create a branch on your fork with these changes and vendor that instead? Thanks again for this!

cc @joelanford @shawn-hurley @theishshah @AlexNPavel @estroz  Thoughts? ^

camilamacedo86 commented:
Hi @LiliC,

Really tks for all your support and help with this subject. No problem at all. Let's waiting for the release 👍

LiliC commented:
@camilamacedo86 FYI there was a new release. I would wait until next week before updating this PR, as I suspect there will be some more changes, but I believe that the maintainer wants to release it for kubecon which is next week. Thanks so much for working on this! https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.2.0-beta.0

yanirq commented:
Hi @camilamacedo86
I am also looking into what is the operator-sdk status with :
> Allow users to watch/cache N namespaces as follows. 

From the operator-sdk point of view I am missing here the part of rbac roles that are created by the operator-sdk for watched namespaces.
Maybe this is out of this PR's scope,  but what is the behaviour or permissions that should be generated/manually edited for the operator when handling N namespaces ? 
Would it be mandatory to configure the operator only as [cluster-scoped](https://github.com/operator-framework/operator-sdk/blob/master/doc/operator-scope.md#operator-scope) ?

LiliC commented:
@camilamacedo86  Thanks for your hard work on this! I will have a test and play around with this today, and maybe we can start working on the migration guide. cc @joelanford 

In the mean time you have some failing tests, I will retest as the error is a bit odd.

/retest

LiliC commented:
/retest

camilamacedo86 commented:
Hi @yanirq,

Here is not a good place to raise questions. Could you please one an issue/question? Please, feel free to ping me and I will try to help you with.

estroz commented on internal/pkg/scaffold/crd.go:
```diff
@@ -24,7 +24,7 @@ import (
 	"github.com/operator-framework/operator-sdk/internal/util/k8sutil"
 
-	"github.com/ghodss/yaml"
+	"sigs.k8s.io/yaml"
```
Remove any `github.com/ghodss/yaml` updates.

camilamacedo86 replied:
done

joelanford commented on pkg/ansible/proxy/proxy.go:
```diff
@@ -38,6 +41,326 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
+type marshaler interface {
```
@camilamacedo86 It looks like these changes may be the result of some merge conflicts (from https://github.com/operator-framework/operator-sdk/pull/1413).

Did this PR require any changes to `pkg/ansible/proxy/proxy.go`? If so, they should be re-applied to the newly refactored `proxy` package.

camilamacedo86 replied:
Done 👍 good spot.

estroz commented:
It looks like we won't be able to upgrade to `v0.2.0` until OLM upgrades to [`kubernetes-1.14.1`](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/go.mod#L66). I did some testing with these changes and saw the following errors (after fixing a few caused by the updates here):

```console
$ dep ensure
$ make install
# github.com/operator-framework/operator-sdk/vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient
vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient/serviceaccount.go:38:2: not enough arguments to return
vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient/serviceaccount.go:38:10: c.Core undefined (type *Client has no field or method Core)
make: *** [Makefile:46: install] Error 2
```

This can be fixed by using `c.CoreV1()` instead of a full `kubernetes-1.14.1` bump, but we still need to get that merged before this works. I'd rather see a bump than this OLM patch anyways.

estroz commented:
OLM PR to bump kube deps: https://github.com/operator-framework/operator-lifecycle-manager/pull/864

estroz commented:
@camilamacedo86 #1512 has all the necessary code changes for the bump. I recommend merging https://github.com/estroz/operator-sdk/tree/bump-controller-runtime-v0.2.0 with your branch and pushing to this PR.

estroz commented on internal/pkg/scaffold/helm/go_mod.go:
```diff
@@ -149,15 +149,15 @@ require (
-// Pinned to kubernetes-1.13.1
+// Pinned to kubernetes-1.14.1
 replace (
 	k8s.io/api => k8s.io/api v0.0.0-20181213150558-05914d821849
```
All of these need updating. You can simply replace the versions with `kubernetes-1.14.1`.

camilamacedo86 replied:
Done

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -8,6 +8,24 @@
+
+### Break changes
+
+- The package `"sigs.k8s.io/controller-runtime/pkg/runtime/scheme"` was deprecated. Replace this import into `pkg/apis/{{projectName}}/v1alpha1/register.go` for `"sigs.k8s.io/controller-runtime/pkg/scheme"` 
```
Done.

camilamacedo86 commented on CHANGELOG.md:
````diff
@@ -8,6 +8,24 @@
+
+```go
+// Replace
+    err = r.client.List(context.TODO(), listOps, podList)
````
Done

camilamacedo86 commented on CHANGELOG.md:
````diff
@@ -8,6 +8,24 @@
+```go
+// Replace
+    err = r.client.List(context.TODO(), listOps, podList)
+    // For
````
Done

camilamacedo86 commented on Gopkg.toml:
```diff
@@ -50,20 +50,21 @@
 [[constraint]]
   name = "github.com/operator-framework/operator-lifecycle-manager"
-  revision = "5eb7ae5bdb7ac36a26d6e34e5d0ec07be128927f"
+  #fixme Waiting for the release 10.1
```
done

camilamacedo86 commented:
Hi @estroz

Really thank you for your help and support here. I rebase the commits in order to make easier we review and go forward as well. Also, I addressed all comments raised by you and I am working on in adding your changes and fixes from #1512 here. I understand that we need both, however, a direct merge/rebase will not work :-( Do you mind I finish it in the weekend?

estroz commented:
@camilamacedo86 #1512 already passes CI, so we should defer to that as the main bump PR. Thanks again for submitting this PR! We appreciate the urgency, as this is a big but necessary change.

camilamacedo86 commented:
Hi @estroz if you could add there what has hee and not there it is great !!.
 had few issues with the merge. tks
